tool/thingtalk-dataset.js

```diff
@@ -0,0 +1,312 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// A tool to manipulate ThingTalk datasets
+//
+// Copyright 2019 National Taiwan University
+//
+// Author: Elvis Yu-Jing Lin <r06922068@ntu.edu.tw> <elvisyjlin@gmail.com>
+//
+// See COPYING for details
+"use strict";
+
+const stream = require('stream');
+const ProgressBar = require('progress');
+
+const ThingTalk = require('thingtalk');
+const Grammar = ThingTalk.Grammar;
+const Library = ThingTalk.Ast.Input.Library;
+const Dataset = ThingTalk.Ast.Statement.Dataset;
+const TokenizerService = require('../lib/tokenizer');
+const OpenCC = require('opencc');
+
+
+class ThingTalkDatasetCleaner extends stream.Transform {
+    constructor(options) {
+        super({ objectMode: true });
+        
+        if (options.keepKeys && options.dropKeys)
+            throw 'keepKeys and dropKeys cannnot be set at the same time.'
+        this.keepKeys = options.keepKeys;
+        this.dropKeys = options.dropKeys;
+        this.options = options;
+        
+        this._defaults = {
+            'id': -1,
+            'type': '',
+            'annotations': {},
+            'utterances': [],
+            'preprocessed': []
+        }
+    }
+
+    _transform(ex, encoding, callback) {
+        if (ex.type == 'example') {
+            let dropKeys = this.dropKeys;
+            if(this.keepKeys) {
+                dropKeys = Object.keys(ex.data).filter(key => !this.keepKeys.includes(key));
+            }
+            if (dropKeys)
+                dropKeys.forEach(key => ex.data[key] = this._defaults[key]);
+        }
+        callback(null, ex);
+    }
+
+    _flush(callback) {
+        process.nextTick(callback);
+    }
+}
+
+class ThingTalkDatasetPreprocessor extends stream.Transform {
+    constructor(options) {
+        super({ objectMode: true });
+        
+        this.locale = options.locale;
+        this.options = options;
+        
+        this._language = '';
+        this._tokenizer = TokenizerService.get(process.env.GENIE_USE_TOKENIZER, true);
+        this._num_examples = 0;
+        this._bar = null;
+        this._i = 0;
+        
+        if (this.locale.toLowerCase() == 'zh-tw') {
+            this._opencc_t2s = new OpenCC('t2s.json');
+            this._opencc_s2t = new OpenCC('s2t.json');
+        }
```

Comment:
This code is language specific, hence it should be in lib/i18n/traditional-chinese.js

Also, I'm assuming the same will be needed at inference time (i.e. when actually taking a sentence from the user)?
If so, it needs to be public API of Genie. Or ideally, it would be in almond-tokenizer.

Reply:
Yes, we also need it when a user type text in Traditional Chinese. However, the Almond tokenizer API need to be changed since it takes "zh" as Chinese, regardless of the Simplified one or the Traditional one.

Reply:
Adding support for country codes in locale tags should be easy to do in almond-tokenizer, and you can then wrap the client-side code appropriately to do the right split, or just pass the unmodified locale tag to almond-tokenizer.